packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,319 @@
+              disabled={this.disabled}
+              required={this.required}
+              ref={(el) => this.pinCodeElements.push(el)}
+            />
+          ))}
+        </div>
+        {hasMessage(this.host, this.message, this.state) && (
+          <StateMessage
+            id="state-message"
+            state={this.state}
+            message={this.message}
+            theme={this.theme}
+            host={this.host}
+          />
+        )}
+      </Host>
+    );
+  }
+
+  private onInput = (
+    e: InputEvent & {
+      target: HTMLInputElement;
+    }
+  ): void => {
+    // needed to update value on auto-complete via keyboard suggestion
+    const { target } = e;
+    if (target.value?.length >= this.length) {
+      const sanitisedValue = getSanitisedValue(target.value, this.length);
+      this.value = sanitisedValue;
+      this.emitUpdateEvent();
+      this.focusFirstEmptyOrLastElement(sanitisedValue);
+    }
+  };
+
+  private onKeyDown = (
+    e: KeyboardEvent & {
+      target: HTMLInputElement & { previousElementSibling: HTMLInputElement; nextElementSibling: HTMLInputElement };
+    }
+  ): void => {
+    const {
+      key,
+      target,
+      target: { previousElementSibling, nextElementSibling },
+    } = e;
+    // prevent default for disabled or loading, but do not impede tab key
+    if (isDisabledOrLoading(this.disabled, this.loading) && key !== 'Tab') {
+      e.preventDefault();
+    } // if input is valid overwrite old value
+    else if (isInputSingleDigit(key)) {
+      e.preventDefault();
+      target.value = key;
+      this.value = getInputValue(this.pinCodeElements);
+      this.emitUpdateEvent();
+
+      if (nextElementSibling) {
+        nextElementSibling.focus();
+      }
+    } // handle alphanumeric keys, allow copy/paste shortcut
+    else if (key?.length === 1 && !(e.ctrlKey || e.metaKey)) {
+      e.preventDefault();
+    } // handle backspace and delete
+    else if (key === 'Backspace' || key === 'Delete') {
+      // transfer focus backward/forward, if the input value is empty
+      if (!target.value) {
+        e.preventDefault();
+        if (key === 'Backspace' && previousElementSibling) {
+          previousElementSibling.value = '';
+          previousElementSibling.focus();
+        } else if (key === 'Delete' && nextElementSibling) {
+          nextElementSibling.value = '';
+          nextElementSibling.focus();
+        }
+      }
+      target.value = '';
+      this.pinCodeElements.map((pinCodeElement) => {
+        if (pinCodeElement === e.target) {
+          pinCodeElement.value = '';
+        }
+        return pinCodeElement;
+      });
+
+      this.value = getInputValue(this.pinCodeElements);
+      this.emitUpdateEvent();
+    } // support native submit behavior
+    else if (key === 'Enter') {
+      if (
+        isWithinForm &&
+        (this.form.querySelectorAll('input').length === 1 || this.form.querySelector('button[type=submit]'))
```

Comment:
What about other form elements like `select` and `textarea`, see https://www.w3schools.com/html/html_form_elements.asp
And how would this work with our `p-button` and `p-button-pure` of they were of `type="submit"`?